File: src/runtime/typescript/src/provider-handlers/generic-handler.ts
/**
 * Generic provider handler for unknown/unsupported vendors.
 *
 * Provides sensible defaults using prompt-based approach similar to Claude.
 *
 * Based on Python's GenericHandler:
 * src/runtime/python/_mcp_mesh/engine/provider_handlers/generic_handler.py
 */

import type { LlmMessage } from "../types.js";
import { formatSystemPrompt as coreFormatSystemPrompt } from "@mcpmesh/core";
import {
  hasMediaParams,
  type ProviderHandler,
  type VendorCapabilities,
  type ToolSchema,
  type OutputSchema,
  type PreparedRequest,
  type OutputMode,
} from "./provider-handler.js";
import { ProviderHandlerRegistry } from "./provider-handler-registry.js";

/**
 * Generic provider handler for vendors without specific handlers.
 *
 * This handler provides a safe, conservative approach that should work
 * with most LLM providers that follow OpenAI-compatible APIs:
 * - Uses prompt-based JSON instructions
 * - Standard tool calling format (via Vercel AI SDK normalization)
 * - No vendor-specific features
 * - Maximum compatibility
 *
 * Use Cases:
 * - Fallback for unknown vendors
 * - New providers before dedicated handler is created
 * - Testing with custom/local models
 * - Providers like: Cohere, Together, Replicate, Ollama, etc.
 *
 * Strategy:
 * - Conservative, prompt-based approach
 * - Relies on Vercel AI SDK to normalize vendor differences
 * - Works with any provider that Vercel AI SDK supports
 */
export class GenericHandler implements ProviderHandler {
  readonly vendor: string;

  constructor(vendor: string = "unknown") {
    this.vendor = vendor;
  }

  /**
   * Prepare request with standard parameters.
   *
   * Generic Strategy:
   * - Use standard message format
   * - Include tools if provided (Vercel AI SDK will normalize)
   * - No vendor-specific parameters
   * - Let Vercel AI SDK handle vendor differences
   */
  prepareRequest(
    messages: LlmMessage[],
    tools: ToolSchema[] | null,
    _outputSchema: OutputSchema | null, // unused in generic handler, but required by interface
    options?: {
      outputMode?: OutputMode;
      temperature?: number;
      maxOutputTokens?: number;
      topP?: number;
      [key: string]: unknown;
    }
  ): PreparedRequest {
    const { outputMode, temperature, maxOutputTokens: maxTokens, topP, ...rest } = options ?? {};

    const request: PreparedRequest = {
      messages: [...messages],
      ...rest,
    };

    // Add tools if provided (Vercel AI SDK will convert to vendor format)
    if (tools && tools.length > 0) {
      request.tools = tools;
    }

    // Add standard parameters if provided
    if (temperature !== undefined) {
      request.temperature = temperature;
    }
    if (maxTokens !== undefined) {
      request.maxOutputTokens = maxTokens;
    }
    if (topP !== undefined) {
      request.topP = topP;
    }

    // Don't add responseFormat - not all vendors support it
    // Rely on prompt-based JSON instructions instead

    return request;
  }

  /**
   * Format system prompt with explicit JSON instructions.
   * Delegates to Rust core.
   *
   * Generic Strategy:
   * - Use detailed prompt instructions (works with most models)
   * - Explicit JSON schema (since we can't assume response_format)
   * - Clear tool calling guidelines
   * - Maximum explicitness for compatibility
   * - Skip JSON schema for text mode (string return type)
   */
  formatSystemPrompt(
    basePrompt: string,
    toolSchemas: ToolSchema[] | null,
    outputSchema: OutputSchema | null,
    outputMode?: OutputMode
  ): string {
    const mode = this.determineOutputMode(outputSchema, outputMode);
    return coreFormatSystemPrompt(
      this.vendor,
      basePrompt,
      !!toolSchemas && toolSchemas.length > 0,
      hasMediaParams(toolSchemas),
      outputSchema ? JSON.stringify(outputSchema.schema) : undefined,
      outputSchema?.name ?? undefined,
      mode,
    );
  }

  /**
   * Return conservative capability flags.
   *
   * For generic handler, we assume minimal capabilities
   * to ensure maximum compatibility.
   */
  getCapabilities(): VendorCapabilities {
    return {
      nativeToolCalling: true, // Most modern LLMs support this via Vercel AI SDK
      structuredOutput: false, // Can't assume all vendors support response_format
      streaming: false, // Conservative - not all vendors support streaming
      vision: false, // Conservative - not all models support vision
      jsonMode: false, // Conservative - use prompt-based JSON instead
    };
  }

  /**
   * Determine output mode - generic handler always uses "hint" for schemas.
   *
   * Since we can't assume structured output support, we rely on
   * prompt-based instructions (hint mode) for JSON schema compliance.
   */
  determineOutputMode(
    outputSchema: OutputSchema | null,
    overrideMode?: OutputMode
  ): OutputMode {
    // Allow explicit override
    if (overrideMode) {
      return overrideMode;
    }

    // No schema means text mode
    if (!outputSchema) {
      return "text";
    }

    // Generic handler uses hint mode (prompt-based JSON instructions)
    // since we can't assume structured output support
    return "hint";
  }
}

// Register as fallback handler
ProviderHandlerRegistry.setFallbackHandler(GenericHandler);

File: src/runtime/python/_mcp_mesh/engine/provider_handlers/generic_handler.py
"""
Generic provider handler for unknown/unsupported vendors.

Provides sensible defaults using prompt-based approach similar to Claude.
"""

import json
import logging
from typing import Any, Optional

from pydantic import BaseModel

from .base_provider_handler import (
    BaseProviderHandler,
    has_media_params,
)

logger = logging.getLogger(__name__)


class GenericHandler(BaseProviderHandler):
    """
    Generic provider handler for vendors without specific handlers.

    This handler provides a safe, conservative approach that should work
    with most LLM providers that follow OpenAI-compatible APIs:
    - Uses prompt-based JSON instructions (like Claude)
    - Standard tool calling format (via LiteLLM normalization)
    - No vendor-specific features
    - Maximum compatibility

    Use Cases:
    - Fallback for unknown vendors
    - New providers before dedicated handler is created
    - Testing with custom/local models
    - Providers like: Cohere, Together, Replicate, Ollama, etc.

    Strategy:
    - Conservative, prompt-based approach
    - Relies on LiteLLM to normalize vendor differences
    - Works with any provider that LiteLLM supports
    """

    def __init__(self, vendor: str = "unknown"):
        """
        Initialize generic handler.

        Args:
            vendor: Vendor name (e.g., "cohere", "together", "unknown")
        """
        super().__init__(vendor=vendor)

    def prepare_request(
        self,
        messages: list[dict[str, Any]],
        tools: Optional[list[dict[str, Any]]],
        output_type: type[BaseModel],
        **kwargs: Any,
    ) -> dict[str, Any]:
        """
        Prepare request with standard parameters.

        Generic Strategy:
        - Use standard message format
        - Include tools if provided (LiteLLM will normalize)
        - No vendor-specific parameters
        - Let LiteLLM handle vendor differences

        Args:
            messages: List of message dicts
            tools: Optional list of tool schemas
            output_type: Pydantic model for response
            **kwargs: Additional model parameters

        Returns:
            Dictionary of standard parameters for litellm.completion()
        """
        request_params = {
            "messages": messages,
            **kwargs,
        }

        # Add tools if provided (LiteLLM will convert to vendor format)
        if tools:
            request_params["tools"] = tools

        # Don't add response_format - not all vendors support it
        # Rely on prompt-based JSON instructions instead

        return request_params

    def format_system_prompt(
        self,
        base_prompt: str,
        tool_schemas: Optional[list[dict[str, Any]]],
        output_type: type,
    ) -> str:
        """
        Format system prompt with explicit JSON instructions.

        Delegates to Rust core for prompt construction.

        Args:
            base_prompt: Base system prompt
            tool_schemas: Optional tool schemas
            output_type: Expected response type (str or Pydantic model)

        Returns:
            Formatted system prompt with explicit instructions
        """
        import mcp_mesh_core

        is_string = output_type is str
        output_mode = "text" if is_string else "hint"

        schema_json = None
        schema_name = None
        if (
            not is_string
            and isinstance(output_type, type)
            and issubclass(output_type, BaseModel)
        ):
            schema_json = json.dumps(output_type.model_json_schema())
            schema_name = output_type.__name__

        return mcp_mesh_core.format_system_prompt_py(
            self.vendor,
            base_prompt,
            bool(tool_schemas),
            has_media_params(tool_schemas),
            schema_json,
            schema_name,
            output_mode,
        )

    def get_vendor_capabilities(self) -> dict[str, bool]:
        """
        Return conservative capability flags.

        For generic handler, we assume minimal capabilities
        to ensure maximum compatibility.

        Returns:
            Conservative capability flags
        """
        return {
            "native_tool_calling": True,  # Most modern LLMs support this via LiteLLM
            "structured_output": False,  # Can't assume all vendors support response_format
            "streaming": False,  # Conservative - not all vendors support streaming
            "vision": False,  # Conservative - not all models support vision
            "json_mode": False,  # Conservative - use prompt-based JSON instead
        }

    def apply_structured_output(
        self,
        output_schema: dict[str, Any],
        output_type_name: Optional[str],
        model_params: dict[str, Any],
    ) -> dict[str, Any]:
        """
        Apply structured output for generic vendors.

        Generic strategy: Don't use response_format since not all vendors support it.
        The consumer should rely on prompt-based instructions instead.

        Args:
            output_schema: JSON schema dict from consumer
            output_type_name: Name of the output type
            model_params: Current model parameters dict

        Returns:
            Unmodified model_params (no response_format added)
        """
        # Don't add response_format - generic vendors may not support it
        # The consumer's system prompt should include JSON instructions
        logger.debug(
            f"⚠️ Generic handler: skipping response_format for '{output_type_name}' "
            f"(vendor '{self.vendor}' may not support structured output)"
        )
        return model_params
